Sources/Server/Parser.swift

```diff
@@ -1,32 +1,68 @@
 import Foundation
 
 public class Parser {
-    
-   public init(){}
-
-   public func parseRequest(request: String) -> Request {
-        let requestComponents = request.components(separatedBy: " ")
-            if (requestComponents[1].contains("?")) {
-                let parsedPathWithQueries = parsePath(path: requestComponents[1])
-                let queries = parseAllQueries(parsedQuery: parsedPathWithQueries[1])
-                return Request(method: requestComponents[0], path: parsedPathWithQueries[0], queries: queries) 
-            }
-        return Request(method: requestComponents[0], path: requestComponents[1])
-   }
-
-   private func parsePath(path: String) -> Array<String> {
+
+    public init(){}
+
+    public func parseRequest(request: String) -> Request {
+        var (requestComponents, requestHeaders) = parseRequest(request: request)
+        let headers = parseHeaders(requestHeaders: requestHeaders)
+        if (requestComponents[1].contains("?")) {
+            let parsedPathWithQueries = parsePath(path: String(requestComponents[1]))
+            let queries = parseAllQueries(parsedQuery: parsedPathWithQueries[1])
+            return Request(method: String(requestComponents[0]), path: parsedPathWithQueries[0], queries: queries) 
+        } else if (headers.keys.contains("Cookie")) {
+            let cookie = parseCookie(headers: headers)
+            return Request(method: String(requestComponents[0]), path: String(requestComponents[1]), cookie: cookie["type"]!) 
+        } else {
+            return Request(method: String(requestComponents[0]), path: String(requestComponents[1]))
+        }
+    }
+
+    private func parseRequest(request: String) -> (Array<Substring>, Array<String>) {
+        var requestAndHeaders = request.components(separatedBy: "\r\n")
+        let requestComponents = requestAndHeaders[0].split(separator: " ")
+        let requestHeaders = parseNonHeaderComponents(requestWithHeaders: requestAndHeaders)
+        return (requestComponents, requestHeaders)
+    }
+
+    private func parseNonHeaderComponents(requestWithHeaders: Array<String>) -> (Array<String>) {
+        var headers = requestWithHeaders
+        headers.removeFirst(1)
+        headers.removeLast(2) 
```

Comment:
As I mentioned before, find a way to clarify why `removeFirst` and `removeLast` are needed here. Ideally, you would NOT just use a comment. :-)

Reply:
I thought "parseNonHeaderComponents" would be a good enough explanation... back to the drawing board.

Reply:
What are non-header components, though?

Reply:
Perhaps use two methods where `headers.removeFirst(1)`, which removes the HTTP request line, could be `removeHTTPRequest()` and `headers.removeLast(2)`, which removes the two new lines, could be `removeNewLine()`

Reply:
Given this example raw request from one of my projects
```
GET /api/rest/v5/agreements/abc123 HTTP/1.1
Access-Token: foo
Host: api.na1.echosign.com
Connection: close
User-Agent: Paw/3.1.5 (Macintosh; OS X/10.13.3) GCDHTTPRequest


```
what do you mean by "the HTTP request line"?

Reply:
HTTP request line would equal first line: `GET /api/rest/v5/agreements/abc123 HTTP/1.1`. In my mind, which may be completely wrong, I labeled the first line as the actual "request" and the following lines as the "headers"